akka-projection-grpc/src/main/scala/akka/projection/grpc/internal/EventPusherConsumerServiceImpl.scala

```diff
@@ -4,42 +4,153 @@
 import akka.persistence.FilteredPayload
+import akka.persistence.query.typed.EventEnvelope
+import akka.persistence.typed.PublishedEvent
+import akka.persistence.typed.ReplicationId
 import akka.persistence.typed.internal.EventWriter
 import akka.persistence.typed.internal.EventWriterExtension
+import akka.persistence.typed.internal.PublishedEventImpl
+import akka.persistence.typed.internal.ReplicatedEventMetadata
+import akka.persistence.typed.internal.ReplicatedPublishedEventMetaData
 import akka.projection.grpc.consumer.scaladsl.EventProducerPushDestination
 import akka.projection.grpc.internal.proto.ConsumeEventIn
 import akka.projection.grpc.internal.proto.ConsumeEventOut
 import akka.projection.grpc.internal.proto.ConsumerEventAck
 import akka.projection.grpc.internal.proto.ConsumerEventStart
 import akka.projection.grpc.internal.proto.EventConsumerServicePowerApi
+import akka.projection.grpc.replication.scaladsl.ReplicationSettings
 import akka.stream.scaladsl.Source
+import akka.util.Timeout
 import io.grpc.Status
 import org.slf4j.LoggerFactory
+
 import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
 import scala.concurrent.Promise
-
-import akka.persistence.query.typed.EventEnvelope
+import scala.concurrent.duration.DurationInt
 
 /**
  * INTERNAL API
  */
 @InternalApi
 private[akka] object EventPusherConsumerServiceImpl {
+
+  private val log = LoggerFactory.getLogger(getClass)
+
   // See akka.persistence.r2dbc.internal.EnvelopeOrigin, but we don't have a dependency
   // to akka-persistence-r2dbc here
   def fromSnapshot(env: EventEnvelope[_]): Boolean =
     env.source == "SN"
 
+  private case class Destination(
+      eventProducerPushDestination: EventProducerPushDestination,
+      sendEvent: EventEnvelope[_] => Future[Any])
+
+  // FIXME weird reverse-package-tree-dependency,
+  //       possibly worth pull factories into grpc.consumer.internal and grpc.replication.internal
+  //       where the other specifics for each lives
+
+  /**
+   * Projection producer push factory, uses the event writer to pass the events
+   */
+  def apply(eventProducerDestinations: Set[EventProducerPushDestination], preferProtobuf: ProtoAnySerialization.Prefer)(
+      implicit system: ActorSystem[_]): EventPusherConsumerServiceImpl = {
+    val eventWriterExtension = EventWriterExtension(system)
+    val destinationPerStreamId = eventProducerDestinations.map { d =>
+      val writerForJournal = eventWriterExtension.writerForJournal(d.journalPluginId)
+
+      val sendEvent = { (envelope: EventEnvelope[_]) =>
+        writerForJournal.askWithStatus[EventWriter.WriteAck](
+          replyTo =>
+            EventWriter.Write(
+              persistenceId = envelope.persistenceId,
+              sequenceNumber = envelope.sequenceNr,
+              event = envelope.eventOption.getOrElse(FilteredPayload),
+              isSnapshotEvent = fromSnapshot(envelope),
+              fillSequenceNumberGaps = false,
```

Comment:
I think this has changed now, see https://github.com/akka/akka-projection/pull/1076/files#diff-86855b40c19d4919a639a34a604a1803fed492db14ab0fa8feaa6e25e9bded11R152